packages/signers/src/deduplicate-signers.ts

```diff
@@ -0,0 +1,17 @@
+import { Address } from '@solana/addresses';
+
+import { MessageSigner } from './message-signer';
+import { TransactionSigner } from './transaction-signer';
+
+/** Removes all duplicated signers from a provided array by comparing their addresses. */
+export function deduplicateSigners<TSigner extends MessageSigner | TransactionSigner>(
+    signers: readonly TSigner[]
+): readonly TSigner[] {
+    const deduplicated: Record<Address, TSigner> = {};
+    signers.forEach(signer => {
+        if (!deduplicated[signer.address]) {
+            deduplicated[signer.address] = signer;
+        }
```

Comment:
We _could_ but I think this would be a premature optimisation.

Let me step back a bit.

The signing and sending flow implemented here is as follows:
1. Given a transaction with signers in its account metas (and optionally more signers provided manually), extract all these signers in an array.
2. Remove duplicates based on the address. **This is where you think we should have a strategy to figure out which ones to keep**.
3. From these deduplicates signers, categorise them. Composite signers (that implement more than one interface) should be chosen depending on what other signers can do. For instance, say Signer A is a `TransactionPartialSigner & TransactionModifyingSigner` and Signer B is a `TransactionModifyingSigner`. Then Signer A should be used as `TransactionPartialSigner` to [avoid modifying conflicts](https://github.com/solana-labs/solana-web3.js/pull/1792#discussion_r1393111926). However, if Signer B was a `TransactionPartialSigner`, then it would make more sense for Signer A to be used as a `TransactionModifyingSigner` as it would be the only one. **This is where I think it matters to have deterministic invariants**. As per our internal document, these invariants are:
    - `TransactionModifyingSigners` must be signed sequentially before any other type of signing.
    - `TransactionPartialSigners` must be signed in parallel.
    - Only one of the `TransactionSendingSigner` must be used last, if any.
    - Use provided fallback logic to send the transaction if no `TransactionSendingSigner` was provided.
    - When a `TransactionSigner` implements more than one interface, use it as a:
        - `TransactionSendingSigner`, if no other `TransactionSendingSigner` exist.
        - `TransactionModifyingSigner`, if no other `TransactionModifyingSigner` exist.
        - `TransactionPartialSigner`, otherwise.
4. Sign and send as per the invariants above.

Now, the reason why I think it is a premature optimisation to add complex invariants to point 2 is that, 99.9% of the time, we are only **filtering the same signer references**. Because it is likely that the same signer will be used multiple time throughout the transaction.

```ts
const mySigner: TransactionSigner = createMySigner();

const instructions = await Promise.all([
  transferSol({ source: mySigner, destination: alice, amount: sol(1.5) }),
  transferSol({ source: mySigner, destination: bob, amount: sol(1.5) }),
]);

const transaction = pipe(
  createTransaction({ version: 0 }),
  tx => appendTransactionInstructions(instructions, tx),
  // ...
);

const signers = getSignersFromTransaction(transaction);
// ^ [mySigner]
```

If we did have complex invariants when deduplicating signers we would be optimising for the following use-case:

"_I have two different signers for the same address and I want it to sign the first instruction differently than the second instruction._"

Which makes no sense at all.

```ts
const myWalletSigner = createWalletSigner(useWallet());
const myKeyPairSigner = generateKeyPairSigner();

const instructions = await Promise.all([
  transferSol({ source: myKeyPairSigner, destination: alice, amount: sol(1.5) }),
  transferSol({ source: myWalletSigner, destination: bob, amount: sol(1.5) }),
]);

const transaction = pipe(
  createTransaction({ version: 0 }),
  tx => appendTransactionInstructions(instructions, tx),
  // ...
);

const signers = getSignersFromTransaction(transaction);
// ^ ??? Why should we have complex invariant for such a non-sense scenario? Let's just pick one.
```

Now, if you decided to split these instructions into two transactions then, the "two signers for the same address" scenario makes more sense and the current implementation still works as expected.

```ts
const myWalletSigner = createWalletSigner(useWallet());
const myKeyPairSigner = generateKeyPairSigner();

const instructions = await Promise.all([
  transferSol({ source: myKeyPairSigner, destination: alice, amount: sol(1.5) }),
  transferSol({ source: myWalletSigner, destination: bob, amount: sol(1.5) }),
]);

const transactionA = pipe(
  createTransaction({ version: 0 }),
  tx => appendTransactionInstructions(instructions[0], tx),
  // ...
);

const transactionB = pipe(
  createTransaction({ version: 0 }),
  tx => appendTransactionInstructions(instructions[1], tx),
  // ...
);

const signersA = getSignersFromTransaction(transactionA);
// ^ [myKeyPairSigner]

const signersB = getSignersFromTransaction(transactionB);
// ^ [myWalletSigner]
```

Reply:
Thanks for all of that explanation! It's awesome to have it down on paper so everyone can look back at it months from now.

I'm a ‘what can go wrong’ guy. Isn't it actually quite likely that someone might go:

1. Create the start of a transaction, somewhere, maybe save it away.
2. Add instructions to it with a `TransactionSigner`.
3. Time passes
4. Decide, later, how exactly to send it (think cluster switchers, endpoint switchers, etc)
5. Add some final instruction with the `TransactionSendingSigner`
6. Call `signAndSendTransactionWithSigners()` on it

…then be left wondering why it failed (ie. the deduper dropped their `TransactionSendingSigner`)

Reply:
I see your point. It's kind of impossible to **find a heuristic that decides which signer to choose** though. You could argue the same the other way around — i.e. someone starts with a `TransactionSendingSigner`, time passes, adds a `TransactionPartialSigner` in another instruction and is left wondering why its own sending logic wasn't used.

I do agree that we could probably warn or even fail here to prevent that from happening in the first place. Meaning, the `deduplicate` function would **only remove duplicated references of signers**. If it finds two different signers object for the same `address` attribute, then it would fail and say something like: "we've identified two different signers for the same account".

What do you think?

Reply:
Resolving this as a decision was made [here](https://github.com/solana-labs/solana-web3.js/pull/1792#issuecomment-1820487421).